test/python/synthesis/test_multi_controlled_synthesis.py

```diff
@@ -232,6 +272,46 @@ def test_mcx_1_clean_b95_cx_count(self, num_ctrl_qubits: int):
         # The bound from the documentation of synth_mcx_1_clean_b95
         self.assertLessEqual(cx_count, 16 * num_ctrl_qubits - 8)
 
+    @data(3, 5, 10, 15)
+    def test_mcx_1_clean_kg24_cx_count(self, num_ctrl_qubits: int):
+        """Test synth_mcx_1_clean_kg24 bound on CX count."""
+        synthesized_circuit = synth_mcx_1_clean_kg24(num_ctrl_qubits)
+        transpiled_circuit = self.pm.run(synthesized_circuit)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # Based on the bound from the Sec 5.1 of arXiv:2407.17966, assuming Toffoli decomposition
+        # requires 6 CX gates.
+        self.assertLessEqual(cx_count, 12 * num_ctrl_qubits - 18)
+
+    @data(3, 5, 10, 15)
+    def test_mcx_1_dirty_kg24_cx_count(self, num_ctrl_qubits: int):
+        """Test synth_mcx_1_dirty_kg24 bound on CX count."""
+        synthesized_circuit = synth_mcx_1_dirty_kg24(num_ctrl_qubits)
+        transpiled_circuit = self.pm.run(synthesized_circuit)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        ## Based on the bound from the Sec 5.3 of arXiv:2407.17966, assuming Toffoli decomposition
+        # requires 6 CX gates.
+        self.assertLessEqual(cx_count, 24 * num_ctrl_qubits - 48)
+
+    @data(3, 5, 10, 15)
+    def test_mcx_2_clean_kg24_cx_count(self, num_ctrl_qubits: int):
+        """Test synth_mcx_2_clean_kg24 bound on CX count."""
+        synthesized_circuit = synth_mcx_2_clean_kg24(num_ctrl_qubits)
+        transpiled_circuit = self.pm.run(synthesized_circuit)
+        cx_count = transpiled_circuit.count_ops()["cx"]
+        # Based on the bound from the Sec 5.2 of arXiv:2407.17966, assuming Toffoli decomposition
+        # requires 6 CX gates.
+        self.assertLessEqual(cx_count, 12 * num_ctrl_qubits - 18)
```

Comment:
I have not read the paper but it seems (both from the assertions here and from manually running the code) that the numbers of CX gates produced by ``synth_mcx_1_clean_kg24`` and by ``synth_mcx_2_clean_kg24`` methods are exactly the same (with the similar statement true for ``dirty``). What is the advantage of using ``2_clean`` method over ``1_clean`` and the other way around?

Reply:
A long time ago I created this rough plot (using optimization_level = 2)
![image](https://github.com/user-attachments/assets/670ffaf5-c054-48a7-bfc9-084110bae884)